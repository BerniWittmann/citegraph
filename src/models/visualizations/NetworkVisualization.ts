import Visualization, { VisualizationBaseFields } from './Visualization'
import NetworkVisualizationTransformer from '@/transformers/NetworkVisualizationTransformer'

export interface NetworkVisualizationParameters {
  pruneLeastCitedPercentage: number
  bibliographicCouplingWeight: number
}

export interface NetworkVisualizationDataNode {
  id: string
  cluster: number
  x: number
  y: number
}

export interface NetworkVisualizationData {
  data: {
    nodes: Array<NetworkVisualizationDataNode>
  }
}

export interface NetworkVisualizationFields {
  data?: NetworkVisualizationData
}

export default class NetworkVisualization extends Visualization implements NetworkVisualizationFields {
  static key: string = 'network'
  key: string = 'network'
  static title: string = 'visualizations.information.network.title'
  static description: string = 'visualizations.information.network.description'
  static longDescription: string = 'visualizations.information.network.long_description'
  static imageUrl: string = 'visualization_images/network.jpg'
  static transformer = NetworkVisualizationTransformer
  parameters: NetworkVisualizationParameters
  data?: NetworkVisualizationData

  constructor ({ id, name, progress, data, timePeriods }: VisualizationBaseFields & NetworkVisualizationFields) {
    super({ id, name, progress, timePeriods })
    this.parameters = {
      pruneLeastCitedPercentage: 5,
      bibliographicCouplingWeight: 50
    }
    this.data = data
  }
}
