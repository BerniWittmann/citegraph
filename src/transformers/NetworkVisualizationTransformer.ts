/* eslint-disable camelcase */
import Transformer from './Transformer'
import { VisualizationBaseFieldsDTO } from '@/models/visualizations/Visualization'
import NetworkVisualization from '@/models/visualizations/NetworkVisualization'

export interface NetworkVisualizationDataNodeDTO {
  id: string
  cluster: number
  x: number
  y: number
}

export interface NetworkVisualizationDTO extends VisualizationBaseFieldsDTO {
  data?: Array<NetworkVisualizationDataNodeDTO>
}

export default class NetworkVisualizationTransformer extends Transformer {
  static fetch (item: NetworkVisualizationDTO): NetworkVisualization {
    return new NetworkVisualization({
      id: item.id,
      name: item.name,
      progress: item.progress,
      timePeriods: item.time_periods,
      data: item.data ? {
        data: {
          nodes: item.data
        }
      } : undefined
    })
  }

  static send (item: NetworkVisualization): NetworkVisualizationDTO {
    return {
      id: item.id,
      name: item.name,
      progress: item.progress,
      key: item.key,
      time_periods: item.timePeriods,
      data: item.data ? item.data.data.nodes : undefined
    }
  }
}
