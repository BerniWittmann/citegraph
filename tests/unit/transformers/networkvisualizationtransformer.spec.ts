import NetworkVisualizationTransformer from '@/transformers/NetworkVisualizationTransformer'
import NetworkVisualization from '@/models/visualizations/NetworkVisualization'

describe('transformers/NetworkVisualizationTransformer', () => {
  describe('can send a NetworkVisualization', () => {
    it('sends an NetworkVisualization', () => {
      expect(NetworkVisualizationTransformer.send(new NetworkVisualization({ id: '1', name: 'First', progress: 0.1 }))).toMatchSnapshot()
    })

    it('sends the data', () => {
      expect(NetworkVisualizationTransformer.send(new NetworkVisualization({ id: '1', name: 'First', progress: 0.1, data: { data: { nodes: [] } } }))).toMatchSnapshot()
    })
  })

  describe('can fetch a NetworkVisualization', () => {
    it('fetches a NetworkVisualization', () => {
      expect(NetworkVisualizationTransformer.fetch({ id: '1', name: 'First', progress: 0.1, key: 'network' })).toMatchSnapshot()
    })

    it('fetches the data', () => {
      expect(NetworkVisualizationTransformer.fetch({ id: '1', name: 'First', progress: 0.1, key: 'network', data: [] })).toMatchSnapshot()
    })
  })
})
