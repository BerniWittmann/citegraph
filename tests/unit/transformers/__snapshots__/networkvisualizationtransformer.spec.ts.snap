// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformers/NetworkVisualizationTransformer can fetch a NetworkVisualization fetches a NetworkVisualization 1`] = `
NetworkVisualization {
  "data": undefined,
  "id": "1",
  "key": "network",
  "name": "First",
  "parameters": Object {
    "bibliographicCouplingWeight": 50,
    "pruneLeastCitedPercentage": 5,
  },
  "progress": 0.1,
  "timePeriods": Array [],
}
`;

exports[`transformers/NetworkVisualizationTransformer can fetch a NetworkVisualization fetches the data 1`] = `
NetworkVisualization {
  "data": Object {
    "data": Object {
      "nodes": Array [],
    },
  },
  "id": "1",
  "key": "network",
  "name": "First",
  "parameters": Object {
    "bibliographicCouplingWeight": 50,
    "pruneLeastCitedPercentage": 5,
  },
  "progress": 0.1,
  "timePeriods": Array [],
}
`;

exports[`transformers/NetworkVisualizationTransformer can send a NetworkVisualization sends an NetworkVisualization 1`] = `
Object {
  "data": undefined,
  "id": "1",
  "key": "network",
  "name": "First",
  "progress": 0.1,
  "time_periods": Array [],
}
`;

exports[`transformers/NetworkVisualizationTransformer can send a NetworkVisualization sends the data 1`] = `
Object {
  "data": Array [],
  "id": "1",
  "key": "network",
  "name": "First",
  "progress": 0.1,
  "time_periods": Array [],
}
`;
